packages/clay-multi-select/src/MultiSelect.tsx

```diff
@@ -0,0 +1,196 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import classNames from 'classnames';
+import ClayIcon from '@clayui/icon';
+import ClayLabel from '@clayui/label';
+import React from 'react';
+
+const COMMA = 188;
+
+const KEYS = [COMMA];
+
+export interface IProps extends React.HTMLAttributes<HTMLInputElement> {
+	children?: React.ReactChild | React.ReactChild[];
+	errorMessage?: string;
+	forwardRef?: React.Ref<HTMLDivElement>;
+	helpText?: string;
+	inputValue?: string;
+	items: string[];
+	label?: React.ReactText;
+	onItemsChange: (val: string[]) => void;
+	onInputChange?: (val: string) => void;
+	spritemap?: string;
+	validateOnBlur?: boolean;
+	validationFn?: (val?: string) => boolean;
+}
+
+const MultiSelect: React.FunctionComponent<IProps> = ({
+	children,
+	errorMessage,
+	forwardRef,
+	helpText,
+	items,
+	label,
+	onItemsChange,
+	spritemap,
+	validateOnBlur = false,
+	validationFn = () => true,
+	inputValue = '',
+	onInputChange,
+	...otherProps
+}) => {
+	const [isValid, setIsValid] = React.useState(true);
+	const [isFocused, setIsFocused] = React.useState();
+
+	const setNewValue = (newVal: string) => {
+		if (validationFn(newVal)) {
+			onItemsChange([...items, newVal]);
+
+			if (onInputChange) {
+				onInputChange('');
+			}
+		} else {
+			setIsValid(false);
+		}
+	};
+
+	const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+		const {keyCode} = event;
+
+		if (inputValue && KEYS.includes(keyCode)) {
+			event.preventDefault();
+
+			setNewValue(inputValue);
+		} else {
+			setIsValid(true);
+		}
+	};
+
+	const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
+		const {value} = event.target;
+
+		setIsFocused(false);
+
+		if (value && validateOnBlur) {
+			setNewValue(value);
+		}
+	};
+
+	const handlePaste = (event: React.ClipboardEvent<HTMLInputElement>) => {
+		const pastedText = event.clipboardData.getData('Text');
+
+		const pastedItems = pastedText
+			.split(',')
+			.map(item => item.trim())
+			.filter(item => item.length);
+
+		if (pastedItems.length > 1) {
+			event.preventDefault();
+
+			const validItems = pastedItems.filter(
+				item => item && validationFn(item)
+			);
+
+			if (validItems.length) {
+				onItemsChange([...items, ...validItems]);
+
+				if (validItems.length === items.length) {
+					setIsValid(false);
+				}
+			}
+		}
+	};
+
+	return (
+		<div className="form-group" ref={forwardRef}>
```

Comment:
@bryceosterhaus I've been thinking we could do a bit differently, instead of having all the components encapsulated here without the outside world being able to tweak it without an API, we can do what we were doing with `Autocomplete` and `DropDown`, that means we will write this component on a lower level.

One thought about this I have is:

- **Validation** can we create any integration with Formik in any way?
- The validation components, we can take to `<ClayForm />`
- `<label>{label}</label>` can we have a `ClayForm.Label`?
- `<div className="input-group input-group-stacked-sm-down">` can we take it to `Clay Form.InputGroup`?

We have used these classes a lot to handle inputs, I think we can move to `ClayForm` and let people think about creating composites and we can make this component high-level, building with them.

Just a mental sketch:
```jsx
const [items, setItems] = useMultiSelect(['one']);

return (
	<ClayMultiSelect>
		<ClayForm.Group>
			<ClayForm.Label>{label}</ClayForm.Label>
			<ClayForm.InputGroup>
				<ClayForm.FormControl>
					{items.map((item, i) => (
						<ClayLabel
							closeButtonProps={{
								onClick: () => {
									setItems([
									...items.slice(0, i),
										...items.slice(i + 1),
									]);
								},
							}}
							key={i}
							spritemap={spritemap}
						>
							{item}
						</ClayLabel>
					))}
					<ClayForm.Input />
				</ClayForm.FormControl>
				<ClayAutocomplete.DropDown>
					{'...'}
				</ClayAutocomplete.DropDown>
				<ClayForm.Feedback />
			</ClayForm.InputGroup>
		</ClayForm.Group>
	</ClayMultiSelect>
);
```

This would be something close, I think it would have some design challenges yet, we will have to deal with a use case that implies in the markup that is to add the loading indicator... It would be something that someone would write composing with the other components to get to that result, but we can deliver all this with a high-level as well.

Reply:
@matuzalemsteles I was actually just thinking about this the other day. `MultiSelect` is essentially a high-level component of `Input`. What do you think about just getting rid of this package and making it a high level version of Input? I think same sort of logic would likely apply to autocomplete. More so works for multi-select since there is no specific class names for it.

`ClayInput`
`ClayInputWithAutoComplete`
`ClayInputWithMultiSelect`

What do you think?

Reply:
Hmm makes sense, I like that. Sounds like a good idea, I think we can test it.

Reply:
Ill send a separate PR today with that general idea for multiselect